kernel/src/main/java/org/kframework/kompile/Kompile.java

```diff
@@ -484,14 +484,44 @@ public Rule parseRule(CompiledDefinition compiledDef, String contents, Source so
     return definitionParsing.parseRule(compiledDef, contents, source);
   }
 
-  private void checkDefinition(Definition parsedDef, Set<Att.Key> excludedModuleTags) {
+  private void preCompilationChecks(Definition parsedDef, Set<Att.Key> excludedModuleTags) {
     scala.collection.Set<Module> modules = parsedDef.modules();
     Module mainModule = parsedDef.mainModule();
     Option<Module> kModule = parsedDef.getModule("K");
     definitionChecks(stream(modules).collect(Collectors.toSet()));
     structuralChecks(modules, mainModule, kModule, excludedModuleTags);
   }
 
+  /**
+   * Run structural checks that rely on the compilation pipeline having already been run.
+   *
+   * <p>The checks run at this stage are:
+   *
+   * <ul>
+   *   <li>Correctness of sorts: that there are no user-supplied parametric sorts remaining, and
+   *       that all non-terminals have a valid sort assigned.
+   *   <li>Validity of the overload lattice; running these checks requires that all
+   *       compiler-generated subsort productions have been created.
+   * </ul>
+   *
+   * @param def The definition to apply checks to.
+   */
+  private void postCompilationChecks(Definition def) {
+    var modules = def.modules();
+    var mainModule = def.mainModule();
+
+    for (Module m : mutable(modules)) {
+      m.checkSorts();
+    }
+
+    checkOverloads(mainModule);
+
+    if (!errors.isEmpty()) {
+      kem.addAllKException(errors.stream().map(e -> e.exception).collect(Collectors.toList()));
+      throw KEMException.compilerError("Had " + errors.size() + " errors after compilation.");
```

Comment:
Does the pre-kompilation check handle singular vs plural in the error message depending on `errors.size()`?

Reply:
Nope, it's just the same implementation as the pre-kompilation check - the plural for 1 error will be reported in both cases.